include/revng/ABI/ApplyRegisterStateDeductions.h

```diff
@@ -0,0 +1,263 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <array>
+#include <iterator>
+#include <optional>
+#include <span>
+#include <variant>
+
+#include "revng/ADT/STLExtras.h"
+#include "revng/Model/ABI.h"
+#include "revng/Model/Architecture.h"
+#include "revng/Model/Register.h"
+#include "revng/Model/RegisterState.h"
+#include "revng/Support/EnumSwitch.h"
+
+namespace detail {
+
+using NullableRegisterState = std::optional<model::RegisterState::Values>;
+
+struct State {
+  NullableRegisterState IsUsedForPassingArguments = std::nullopt;
+  NullableRegisterState IsUsedForReturningValues = std::nullopt;
+};
+
+template<model::Architecture::Values Architecture>
+class RegisterStateMap {
```

Comment:
I'm a bit skeptical about `detail::RegisterStateMap` and `RegisterStateMap`. It's a significant amount of (non-reusable) complexity to have the `std::array` of the *exact* size but then all the instance (AFAIU) end up in a variant, so at run-time you don't get much benefit.

I'd consider wrapping a `std::map<model::Register::Values, State>` (or some other map-like data structure) that's initially populated with all the things you need and the user is not able to add new entries but only change mapped values.

I think this kind of complexity is justified *only* if it's either easily reusable or brings a significant performance benefit.

Reply:
Depends, if we're still around ~250 lines for defining a data structure which could be rather simple, I'm not very happy.
If it helps cutting out `detail::RegisterStateMap`, it's probably good.

Reply:
Works!